### DISCUSSION TITLE
'there is no reactor running, must be called from the context of a Tokio 1.x runtime'

### DISCUSSION THREAD

**Discussion by nikkolasg** [libp2p/rust-libp2p#2592] [Q&A] [answered]
Hi all,

Very new user to rust-libp2p (and libp2p in general) here, so please bear with me if I'm forgetting something obvious. I've basically took the rendezvous examples, for the [rendezvous point ](https://github.com/libp2p/rust-libp2p/blob/master/protocols/rendezvous/examples/rendezvous_point.rs) and the register+discover node.
I'm using tokio, and running all this inside a `#[tokio::test] rendezvous() -> Result<()> { ...  }` test method.
Running the discover loop seem to work fine but when I run the "discovering" nodes  (inside my test) I get the following:
```
thread 'libp2p-swarm-task-0' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', /.../.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/context.rs:29:26
```
2 notes:
1. I see only one event coming from the swarm where it actually listens, i.e.  `Handled event NewListenAddr { listener_id: ListenerId(1), address: "/ip4/127.0.0.1/tcp/5004" }` and then I see this this error msg.
2.  My code is indeed running inside a tokio runtime, since the test runs, I'm using async everywhere and my `loop { tokio::select { .... } }` is still running as I see the tick appearing (to ask for more peers from rendezvous server). Problem is I have zero events coming from the swarm (except the `NewListenAddr` shown before), which makes sense since apparently it panic'd. 
3. Relevant lines of cargo:
```
libp2p = { version = "^0.43" , default-features = false, features = ["rendezvous","yamux","mplex","noise","websocket","tcp-tokio","dns-tokio","ping","identify"] }
tokio = { version = "1.17", features = ["full"] }
```

Any idea what could be causing this ? 

Thank you :pray:

**Comment by nikkolasg** [comment, marked as answer]
Ok, I found the related issue https://github.com/libp2p/rust-libp2p/issues/2230 and then found how to setup properly the tokio executor [here](https://github.com/libp2p/rust-libp2p/commit/6b3500b931f1807ef80856c8ec67078773cb0063#diff-fcfd14328b2fef8a8b28b0199e0a7ad982da80c45b45e5d88b0b69ed6cb3a99eR135)

For reference, solved it with
```rust
let mut swarm = SwarmBuilder::new(
            tokio_development_transport(self.kp.clone()).unwrap(),
            MyBehaviour { ... }, ... )
        .executor(Box::new(|fut| {
            tokio::spawn(fut);
        }))
        .build();
```

**Reply by michael-herwig** [reply]
I want to add, as with version `0.50` of `libp2p`, the `SwarmBuilder::new` is deprecated and encourages usage of constructors explicitly stating the executor (`SwarmBuilder::with_executor`/`SwarmBuilder::wihtout_executor`). For this particular example, there is a method for `tokio`, namely `SwarmBuilder::with_tokio_executor`. Just in case someone runs into the sample problem as me upgrading libp2p.

**Comment by mxinden** [comment]
@nikkolasg thanks for referencing the solution. Mind proposing a patch improving the docs, preventing future users to run down the same rabbit hole?
